Patterns api

Adadch opened the pull request:
This is just the first modification with a headings small fix to the file.

After build the "built-in patterns" heading is remaining empty and not taking the functions of the python file which was fixed by adding `:members:` inside automodule.

And a warning is coming up for me  "WARNING: html_static_path entry '_static' does not exist" which was fixed by simply adding a _static folder inside source folder.

Adadch commented:
after the last modification this morning i noticed that using the :members: for the automodule did give the results but they are in the wrong order. Order example (class CartesianPattern is first then CategorySetPattern, then ClosedIntervalPattern etc.) 
Along with their functions, none of the functions with __ like (__repr__ and __and__ etc.) are shown in the result. And the order is as well messed up but in a different way. Here the properties are being shown first then it's other functions.

Adadch commented:
does the pattern class also need to have its functions shown using :member: ? since it's written in the docs that it is an abstract class

EgorDudyrev commented:
> does the pattern class also need to have its functions shown using :member: ? since it's written in the docs that it is an abstract class

Yes. The general `Pattern` class should also be explicitly described.
From one side, it is an abstract class, from the other: it is really not. `Pattern` class does not implement all the functions that one can do with more precise patterns: for example, it does not know how to compute atomic patterns. At the same time, it supports some basic operations such as meet and join. So, if you create patterns as `a, b = Pattern(frozenset('ab')), Pattern(frosenset('bc'))`, the meet operation `a & b` and the join operation `a | b` would actually work. If `Pattern` would be a true abstract class, even creating such pattern variables `a` and `b` would be impossible.

Adadch commented:
> Yes. The general Pattern class should also be explicitly described.
From one side, it is an abstract class, from the other: it is really not. Pattern class does not implement all the functions that one can do with more precise patterns: for example, it does not know how to compute atomic patterns. At the same time, it supports some basic operations such as meet and join. So, if you create patterns as a, b = Pattern(frozenset('ab')), Pattern(frosenset('bc')), the meet operation a & b and the join operation a | b would actually work. If Pattern would be a true abstract class, even creating such pattern variables a and b would be impossible.

That is to my understanding of the file, but then why did you call it abstract in the docs? only because it doesn't support or compute all the functions/operations?

Adadch commented:
I added the :members: to the pattern class, but i did notice a new alignement issue with the result, the properties mentioned in the class docstring are written same as the functions and on the same level as headings as well which i don't know how to correct

Adadch commented:
> I added the :members: to the pattern class, but i did notice a new alignement issue with the result, the properties mentioned in the class docstring are written same as the functions and on the same level as headings as well which i don't know how to correct PS like it happened for classes in built-in patterns, the functions here have changed order and now are ordered alphabetically

the only "fix" i managed to get for the alignement issues is to change the max depth of the toctree in the index file to 3, which removes all the written functions in it except for the functions in the algorithms_api file since it has only functions and no classes

As for the ordering of the functions/classes it was an easy fix in the conf.py file, just need to add autodoc_member_order = 'bysource' which is enough to order everything as it is written in the original files instead of alphabetically

EgorDudyrev commented:
> > Yes. The general Pattern class should also be explicitly described.
> > From one side, it is an abstract class, from the other: it is really not. Pattern class does not implement all the functions that one can do with more precise patterns: for example, it does not know how to compute atomic patterns. At the same time, it supports some basic operations such as meet and join. So, if you create patterns as a, b = Pattern(frozenset('ab')), Pattern(frosenset('bc')), the meet operation a & b and the join operation a | b would actually work. If Pattern would be a true abstract class, even creating such pattern variables a and b would be impossible.
> 
> That is to my understanding of the file, but then why did you call it abstract in the docs? only because it doesn't support or compute all the functions/operations?

I remember I was saying something about abstract classes, but I cannot find any precise citation neither in the code (including the commits back from February) nor in the pull requests.
I think I was talking about Pattern as of "somewhat abstract" class, not the actual abstract class. Because `Pattern` class is intended to be mostly used as a top-level abstract class to inherit other classes from, but, at the same time, it also supports some basic functionality.

Adadch commented:
the latest commit has a few changes since it holds the modifications i've done to both conf.py and index.rst as well.

for **index.rst**: the added maxdepth for the toctree since i couldn't find any way to organize it any better in the patterns_api file.
PS: i think we can delete the other toctree as well since it serves nothing in its current state.

for **conf.py**: added autodoc_member_order = 'bysource' line to make the order of :members: go by the order of the source files and not alphabetically.

for the main file **patterns_api.rst**: added a few extra empty lines and a dotted line to make the functions separators go as heading 4 instead of heading 3 so they don't appear in the toctree after build anymore.

EgorDudyrev commented:
> the latest commit has a few changes since it holds the modifications i've done to both conf.py and index.rst as well.
> 
> for **index.rst**: the added maxdepth for the toctree since i couldn't find any way to organize it any better in the patterns_api file. PS: i think we can delete the other toctree as well since it serves nothing in its current state.
> 
> for **conf.py**: added autodoc_member_order = 'bysource' line to make the order of :members: go by the order of the source files and not alphabetically.
> 
> for the main file **patterns_api.rst**: added a few extra empty lines and a dotted line to make the functions separators go as heading 4 instead of heading 3 so they don't appear in the toctree after build anymore.
> 
> Edit: the dotted line in patterns_api isn't doing anything for the build so I'll either have to put an actual heading 3 above the functions separators or leave them be in the toctree as they are now. meaning delete the dotted line and change the commas lines into dotted ones

Thank you. The index file looks fine for me. I would also like to make visual and fancy later on. But that is a whole other story.

Please resolve the two problems I have mentioned earlier and then we can close this Pull Request.

EgorDudyrev commented on docs/source/patterns_api.rst:
```diff
@@ -1,19 +1,96 @@
 Patterns API
 ============
 
```
I have noticed that now we have all the section titles mentioned on the side of the page in "On this page" section.
However, in many other documentations, most of the sections are mentioned in the left part of the page in the "Section Navigation" menu.

Could you move the section titles from the right side of the page to the left side?
For example, you could look at how it was made in matplotlib documentation https://matplotlib.org/stable/api/colors_api.html , or in numpy documentation https://numpy.org/doc/stable/reference/routines.linalg.html, or in pandas documentation https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html.

What I would like to see with the Patterns API is that the sections "Pattern Class", "Built-in Patterns", "Built-in Patterns/ItemSetPattern", "Built-in Patterns/CategorySetPattern", etc would be shown on the left part of the screen within "Section Navigation" menu. And then, when we go to a specific page (say "Built-in Patterns/ItemSetPattern"), we will subsections "PatternValueType", "Universe", "atomic_patterns", etc on the right part of the screen in the "On this page" menu.

Adadch replied:
So just to be clear, adding a "toctree" on the side of the file as well in which it has all the function classes and groupements together?

EgorDudyrev replied:
I quite like the current version of the page.
One last detail, the left page parts in both numpy and pandas documentation have "drop down" buttons (I am not sure if I got their names right). For example, when you go to Pandas documentation (https://numpy.org/doc/stable/reference/index.html), you see on the left that there is something about "DataFrame". And then you can see "the insides" of DataFrame only if you click on "DataFrame".

Could you add the same buttons Patterns api?
The problem is that, for now, we have too many details about sections and subsections on the left, that it becomes hard to see the general picture of Patterns API having "Patterns", "Built-in Patterns", and "Pattern Structure".
It would be better if we could open the Patterns API, see that there are "Patterns", "Built-in Patterns", and "Pattern Structure", and then we could click on "Built-in Patterns" to see some details on this part of the api.

Adadch replied:
i'll look it up

Adadch replied:
okay so now i understand how pandas did it, not how i can do it. they have folders and not files in their toctree
each folder leads to a different index.rst file which has each choice as a separate page not all on the same page

I'll look around how i can make the drop down for headings in the same page, it'll probably be the same as in "word" from the office group, each lower headings are nested in the upper one

EgorDudyrev replied:
Ok making one new button requires creating a bunch of new files, then we can just leave it for another time.
If it requires adding one more command in "conf.py" file, then please do it in this Pull Request

Adadch commented on docs/source/patterns_api.rst:
```diff
@@ -2,18 +2,93 @@ Patterns API
 
 .. autoclass:: paspailleur.pattern_structures.pattern.Pattern
-
+   :members:
```
I looked around a lot but couldn't find any part that could remove the duplications like this after that i started using `make clean; make html` so the build itself is deleted and recreated each time which somehow did limit the amount of duplications but not entirely.

I'll look further into it anyway.

EgorDudyrev replied:
I had a look at the documentation for the documentation style that we subconsciously follow: https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring .
And it says that class properties should be defined in the "Attributes" section close the attributes of a class.

Could you try to change the docstring of Pattern class and group all properties, public, and private attributes under "Attributes" section. This might help to solve the duplication problem.

Adadch replied:
will do it now and check

Adadch replied:
the docstring now looks like this 
``    Attributes
    ----------
    PatternValueType: 
        A type variable representing the type of the pattern's value.
    _value: 
        The processed value of the pattern.

    value
        Return the value of the pattern.
    meetable
        Check if meet (intersection) operation is defined for this Pattern class.
    joinable
        Check if join (union) operation is defined for this Pattern class.
    atomisable
        Check if the pattern can be atomized.
    etc.``

and the result is this 
PatternValueType
A type variable representing the type of the pattern’s value.

_value
The processed value of the pattern.

value
Return the value of the pattern.

meetable
Check if meet (intersection) operation is defined for this Pattern class.

joinable
Check if join (union) operation is defined for this Pattern class.

atomisable
Check if the pattern can be atomized.

with the functions in their respective sections

Adadch replied:
so it did work by removing all but the word attributs so should i commit the new change?

EgorDudyrev replied:
Yes. But make sure that everything looks nice on the html version.
The numpydoc documentation says that one should not add explanations for properties in class's docstring, if the properties themselves have their own docstrings. I assume if you add additional explanation for properties in the docstring of a class, it might break something in Sphinx compiler

Adadch replied:
so should i remove the properties part completely from the class docstring?

EgorDudyrev replied:
Could you commit the docstring that has "Attributes" section and has no "Properties" section. So I could see the current state and then decide if we remove all this information?

Adadch commented:
With the latest commit I added the toctree to the side which needed another modification to the conf file. 

I deleted the main toctree from index file which only took some extra space with no added value

Added the functions and their sections to the patterns_api file, but still can't find a way to remove the properties and its contents from being considered sections by themselves

Adadch commented:
after all the work i noticed an (error?) in the look of the tree, I even changed its theme and the problem persisted on the right side tree instead of the changed left side.
some parts in the file light up when you reach them while scrolling, others don't and i have no idea why since both the conf.py, the .rst file and the html file all should be working well. i even checked the created html file between two parts one which light up and the other which does not and the code is completely the same excluding the name 

the ones that don't work now are "patterns' merge" from Pattern class. The entire built-in patterns class (could be a different reason for the classes but the heading of the class itself should light up but doesn't).  "Properties that are easy to compute
" from Pattern_structures class. those are the only parts that don't light up in the tree

Adadch commented:
The latest commit is where i changed the docstrings of the files to remove the "properties" from the page tree and to initiate the attributes/ private attributes/ properties to look as headings without being considered headings in the html. 

That was only possible after changing the (----) lines below them in each class into (.....) lines. So I'm not so sure about its validity on term of docstrings
